### PR TITLE
bisect lakectl support

### LINKED ISSUE #5218: lakectl "git-bisect" like capability for fast lookup for commit

### PR DESCRIPTION
**Pull request by nopcoder**
Close #5218

### CODE REVIEW

File: cmd/lakectl/cmd/bisect.go [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+}
+
+//nolint:gochecknoinits
+func init() {
+	rootCmd.AddCommand(bisectCmd)
+
+	bisectCmd.AddCommand(bisectStartCmd)
+	bisectCmd.AddCommand(bisectResetCmd)
+	bisectCmd.AddCommand(bisectBadCmd)
+	bisectCmd.AddCommand(bisectGoodCmd)
```
Comment (itaiad200) [inline cmd/lakectl/cmd/bisect.go:371]: Maybe use different naming?
https://git-scm.com/docs/git-bisect#_alternate_terms

File: cmd/lakectl/cmd/bisect.go [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+
+const (
+	BisectSelectGood BisectSelect = iota
+	BisectSelectBad
+)
+
+var ErrCommitNotFound = errors.New("commit not found")
+
+// bisectCmd represents the bisect command
+var bisectCmd = &cobra.Command{
```
Comment (itaiad200) [inline cmd/lakectl/cmd/bisect.go:47]: I think we should hide it by default. This is very hardcore plumbing..

File: cmd/lakectl/cmd/bisect.go [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+	BadCommit  string        `json:"badCommit,omitempty"`
+	GoodCommit string        `json:"goodCommit,omitempty"`
+	Commits    []*api.Commit `json:"commits,omitempty"`
+}
+
+const bisectStartCmdArgs = 3
+
+// bisectStartCmd represents the start command
+var bisectStartCmd = &cobra.Command{
+	Use:   "start <repository> <bad ref> <good ref>",
```
Comment (itaiad200) [inline cmd/lakectl/cmd/bisect.go:64]: Why do we need the repo if we have the refs?

File: cmd/lakectl/cmd/bisect.go [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+func (b *Bisect) SaveSelect(sel BisectSelect) error {
+	if len(b.Commits) <= 1 {
+		return nil
+	}
+	h := len(b.Commits) >> 1
+	switch sel {
+	case BisectSelectGood:
+		b.Commits = b.Commits[:h]
+	case BisectSelectBad:
+		b.Commits = b.Commits[h:]
```
Comment (itaiad200) [inline cmd/lakectl/cmd/bisect.go:357]: OffByOne bugs seem inevitable, how do we protect against them?

**Review by itaiad200 — CHANGES_REQUESTED** [review]

File: cmd/lakectl/cmd/bisect.go [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+	"github.com/treeverse/lakefs/pkg/api"
+)
+
+const bisectCommitTemplate = `{{ range $val := . }}
+ID:            {{ $val.Id|yellow }}{{if $val.Committer }}
+Author:        {{ $val.Committer }}{{end}}
+Date:          {{ $val.CreationDate|date }}
+Meta Range ID: {{ $val.MetaRangeId }}
+{{ if gt ($val.Parents|len) 1 -}}
+Merge:         {{ $val.Parents|join ", "|bold }}
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply cmd/lakectl/cmd/bisect.go:26]: Present parents if we have more than 1 - like in commit log

File: cmd/lakectl/cmd/bisect.go (hunk 3 above)
Comment (nopcoder) [reply cmd/lakectl/cmd/bisect.go:64]: At first the two refs were optional - fixing.

[the remark this answers is not in the document]
Comment (nopcoder) [reply cmd/lakectl/cmd/bisect.go:64]: removed the repository

File: cmd/lakectl/cmd/bisect.go [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+func (b *Bisect) Update(ctx context.Context, client api.ClientWithResponsesInterface) error {
+	if b.GoodCommit == "" || b.BadCommit == "" {
+		b.Commits = nil
+		return nil
+	}
+
+	// scan commit log from bad to good (not included) and save them into state
+	b.Commits = nil
+	var commits []*api.Commit
+	params := &api.LogCommitsParams{}
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply cmd/lakectl/cmd/bisect.go:325]: Good idea not to use the default - will use 500. Not sure the caller will want to scan all the commit log.
Later we can have it configured / by request.

File: cmd/lakectl/cmd/bisect.go [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+
+		badURI := MustParseRefURI("bad", args[1])
+		if badURI.Repository != repoURI.Repository {
+			Die("Repository doesn't match 'bad' commit", 1)
+		}
+
+		goodURI := MustParseRefURI("good", args[2])
+		if goodURI.Repository != repoURI.Repository {
+			Die("Repository doesn't match 'good' commit", 1)
+		}
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply cmd/lakectl/cmd/bisect.go:79]: yes, branch, tag - I've updated the error message to call use 'reference' - from this point we will print out and use commnits.

File: cmd/lakectl/cmd/bisect.go [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+		fmt.Printf("Good commit lakefs://%s/%s, waiting for bad commit\n", b.Repository, b.GoodCommit)
+	case commitCount == 0:
+		fmt.Println("No commits found")
+	case commitCount == 1:
+		commit := b.Commits[0]
+		fmt.Printf("Found commit lakefs://%s/%s %s\n", b.Repository, commit.Id, commit.Message)
+	default:
+		steps := math.Log2(float64(commitCount))
+		h := (commitCount >> 1) - 1
+		fmt.Printf("Bisecting: %d commits left to test after this (roughly %d steps)\n", h, int(steps))
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply cmd/lakectl/cmd/bisect.go:131]: Can you try the equal `git` command - I know it will sound strange but I think I replicated the same behaviour.

**Comment by itaiad200** [comment]
@nopcoder bad merge/rebase?

**Comment by nopcoder** [comment]
@itaiad200 rebased + esti test

File: cmd/lakectl/cmd/bisect.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (itaiad200) [reply cmd/lakectl/cmd/bisect.go:26]: Why is the label `Merge:` and not `Parent(s):`?

**Review by itaiad200 — APPROVED** [review]
Really cool addition - thanks for adding it

**Comment by nopcoder** [comment]
Used the same syntax of `lakectl log`

**Pull request merged**